woocommerce/google-listings-and-ads
# Connect programs report page to the API data.

### Changes proposed in this Pull Request:

- Add `useProgramsReport` hook, with glue code for aggregating free and paid programs. (b79c9fc)
- Add mocked `(ads|mc)/reports/programs` responses for testing (075cd8d)
- Connect programs `SummarySection` to API data (74b0308)
	- Add optional `expectedLength` prop to `SummarySection`, so it could render applicable placeholder even before the metrics array is loaded/resolved.
- Add and connect programs's `ChartSection` to the API,(695f91d)
	- Make `ChartSection` accept empty `metrics` prop, so it could render a placeholder, when `metrics` data is being loaded/resolved.


Implements part of #242 

### Screenshots:

![Screencast of clicking through program reports connected to data](https://user-images.githubusercontent.com/17435/118124444-e0836900-b3f5-11eb-929a-dd38517cad5e.gif)



### Detailed test instructions:

1. Connect to an account that has some reports data, or use mocked responses, by adding at the end of [`google-listings-and-ads.php`](https://github.com/woocommerce/google-listings-and-ads/blob/trunk/google-listings-and-ads.php#L91) following snippet:
	  ```php
	  
	  add_filter(
		  'gla_prepared_response_ads-reports-programs',
		  function( $response ) {
			  return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/programs.json' ), true ) ?: [];
		  },
	  );
	  add_filter(
		  'gla_prepared_response_mc-reports-programs',
		  function( $response ) {
			  return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/programs.json' ), true ) ?: [];
		  },
	  );
	  ```
2. Go to Programs Report page  [`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms)
3. Click through Summary Numbers and check that Chart updates accordingly,
4. Check that tooltip about missing data is rendered,
5. Switch programs filter value to `Single program->"Free Listings"`
6. Redo 3 & 4
7. Switch programs filter value to `Single program->any "Campaign"`
6. Redo 3 & 4

### Not covered in this PR:
1. Intervals data for the chart is not aggregated, what results with chart line going back https://github.com/woocommerce/google-listings-and-ads/issues/589 
2. Programs Table will be updated in another PR

### Doubts:

1. Currently, until reports data come we render a fancy preloader for all 5 metrics, which may eventually be replaced with only 2 metrics, if 
	1. User selected `Single program -> "Free Listings"`
	2. User selected `All` and does not have any paid campaigns
	3. User selected `All` but paid campaigns do not have any report data yet.
	4. User selected `Compare`, and choose `free listings, campaign#12` but the paid campaign no longer exists.

	I'm not sure if that 100% desired behavior. The alternative solution that comes to my mind is:
	Decide the number of Metrics based on the selection, not the actually available data, and show 
	_i._ - a placeholder for 2 metrics that resolves to 2 metrics with data,
	_ii., iii., iv._ - a placeholder with 5 metrics that resolves to 5 metrics but some with `0` or `N/A`.
	

### Changelog Note:

> Connected Programs Report Summary and Chart to API data.
